feat: add replay syncer and get replay status

arinda-arif opened the pull request.

kushsharma commented on cmd/replay.go:
```diff
@@ -267,3 +270,90 @@ func runReplayRequest(l logger, projectName, namespace, jobName, startDate, endD
+that generated when starting a replay.
+		`,
+		Args: func(cmd *cli.Command, args []string) error {
+			if len(args) < 2 {
```
In the description, you said we need 1 argument, here we are looking for 2.

arinda-arif replied:
updated to only 1

kushsharma commented on cmd/replay.go:
```diff
@@ -267,3 +270,90 @@ func runReplayRequest(l logger, projectName, namespace, jobName, startDate, endD
+		`,
+		Args: func(cmd *cli.Command, args []string) error {
+			if len(args) < 2 {
+				return errors.New("job name and replay ID are required")
```
Btw, isn't replay id unique, why do we need a job name?

arinda-arif replied:
yes, it is unique. have added it for the api call: https://github.com/odpf/proton/blob/4bcadc9389fa7e895509fa85cd7a2c25af5f3dc6/odpf/optimus/runtime_service.proto#L192
let me know your thoughts and if you think removing it is more proper

kushsharma replied:
For rest call to make it follow the convention I guess we need it there but we are using GRPC here, do we really need it?

arinda-arif replied:
agreed. updated to only require replay ID here

kushsharma commented on cmd/server/server.go:
```diff
@@ -314,7 +314,8 @@ func Initialize(conf config.Provider) error {
 		hash: appHash,
 	}
-	registeredProjects, err := projectRepoFac.New().GetAll()
+	projectRepo := projectRepoFac.New()
```
why break this into 2 lines?

arinda-arif replied:
merged into 1

kushsharma commented on job/replay_manager.go:
```diff
@@ -37,7 +40,9 @@ type ReplayManagerConfig struct {
-	Replay(context.Context, *models.ReplayWorkerRequest) (string, error)
+	Replay(context.Context, *models.ReplayRequest) (string, error)
+	GetReplay(uuid uuid.UUID) (*models.ReplaySpec, error)
+	GetRunsStatus(ctx context.Context, reqInput *models.ReplayRequest, jobName string) ([]models.JobStatus, error)
```
I think `GetRunStatus` sounds better

arinda-arif replied:
updated

kushsharma commented on job/replay_manager.go:
```diff
@@ -236,6 +133,18 @@ func (m *Manager) spawnServiceWorker() {
+		cancelCtx()
+	}
+	logger.I("replays synced")
+	cancelCtx()
```
We can `defer` this call before line 140 right to avoid calling it at 2 places.

arinda-arif replied:
changed to use defer

arinda-arif commented on api/handler/v1/runtime.go:
```diff
@@ -803,23 +806,60 @@ func (sv *RuntimeServiceServer) Replay(ctx context.Context, req *pb.ReplayReques
+	replayState, err := sv.jobSvc.GetStatus(ctx, replayRequest)
 	if err != nil {
-		return nil, status.Errorf(codes.NotFound, "%s: namespace %s not found", err.Error(), req.GetNamespace())
+		return nil, err
```
yes, but for this i have set the status.Error inside the called function (getProjectSpec and getJobSpec). let me know if this is make sense or I should only return the error message and wrap the status.Err in the line you mentioned

kushsharma replied:
I don't see any of these functions, I see `replayState, err := sv.jobSvc.GetStatus(ctx, replayRequest)`

arinda-arif replied:
updated

arinda-arif commented on cmd/replay.go:
```diff
@@ -267,3 +270,90 @@ func runReplayRequest(l logger, projectName, namespace, jobName, startDate, endD
+		Example: "optimus replay status replay-id",
+		Long: `
+This operation takes one argument, replay ID[required]
+that generated when starting a replay.
```
got it

arinda-arif replied:
updated

kushsharma commented on job/replay_manager.go:
```diff
@@ -37,7 +40,9 @@ type ReplayManagerConfig struct {
 type ReplayManager interface {
 	Init()
-	Replay(context.Context, *models.ReplayWorkerRequest) (string, error)
+	Replay(context.Context, *models.ReplayRequest) (string, error)
```
Any reason we are using pointer in the request?

arinda-arif replied:
was read somewhere that instead of passing the whole struct, better practice is to reference it. let me know your thoughts & consideration?

kushsharma replied:
If the struct is large then only it makes sense, in these cases, to pass it via reference, else it is actually bad for cache coherence. CPU works with L1, L2, L3... caches, each of them has different sizes, it's actually faster to copy a small struct and keep it in the cache then to dereference a pointer.

arinda-arif replied:
Thank you for the explanation. I have updated to not use pointer in replay requests.

kushsharma commented on job/replay_manager.go:
```diff
@@ -37,7 +40,9 @@ type ReplayManagerConfig struct {
 	Init()
-	Replay(context.Context, *models.ReplayWorkerRequest) (string, error)
+	Replay(context.Context, *models.ReplayRequest) (string, error)
+	GetReplay(uuid uuid.UUID) (*models.ReplaySpec, error)
```
Any reason we are using pointer here as well?

arinda-arif replied:
Have updated this too to not use pointer

kushsharma commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,181 @@
+	projects                  []models.ProjectSpec
+}
+
+func NewReplaySyncer(replaySpecFactory ReplaySpecRepoFactory, scheduler models.SchedulerUnit, dependencyResolver DependencyResolver, projectJobSpecRepoFactory ProjectJobSpecRepoFactory, assetCompiler AssetCompiler, projects []models.ProjectSpec) *Syncer {
```
We can format these lines to have a soft limit to 80 and a hard limit to 120 chars

arinda-arif replied:
i see you have refactored this, thanks!

kushsharma commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+func NewReplaySyncer(replaySpecFactory ReplaySpecRepoFactory, scheduler models.SchedulerUnit, dependencyResolver DependencyResolver, projectJobSpecRepoFactory ProjectJobSpecRepoFactory, assetCompiler AssetCompiler, projects []models.ProjectSpec) *Syncer {
+	return &Syncer{replaySpecFactory: replaySpecFactory, scheduler: scheduler, dependencyResolver: dependencyResolver, projectJobSpecRepoFactory: projectJobSpecRepoFactory, assetCompiler: assetCompiler, Now: time.Now, projects: projects}
```
time should be injected in the constructor and with UTC

arinda-arif replied:
do this mean 'Now' value should also passed to the constructor, not being initialized in the constructor?

arinda-arif replied:
done by Kush

kushsharma commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+func (s Syncer) Sync(context context.Context, runTimeout time.Duration) error {
+	replaySpecRepo := s.replaySpecFactory.New(models.JobSpec{})
```
Why do we need to pass this empty jobspec?

arinda-arif replied:
initially we are constructing ReplayRepository with a JobSpec, but just realized now that we don't really need it because of the latest changes. will update

arinda-arif replied:
updated. not passing jobspec to constructor

kushsharma commented on models/job.go:
```diff
@@ -58,6 +58,7 @@ type JobSpec struct {
 	Dependencies map[string]JobSpecDependency // job name to dependency
 	Assets       JobAssets
 	Hooks        []JobSpecHook
+	Project      ProjectSpec
```
Why is this added?

arinda-arif replied:
have added this because when getting Replay from the repository, need to have Job and Project. second thought, i will remove this and use project from user input instead

arinda-arif replied:
another usage: on the replay syncer, we need to know the project of each active replays that need to be synced, so in the replay repository, we fetch Replay that got preloaded with Job and Project (that is why Project is added). Thoughts?

arinda-arif replied:
yes, i can remove this and call project where it is needed (when calling scheduler)

kushsharma commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,181 @@
+	projectJobSpecRepoFactory ProjectJobSpecRepoFactory
+	assetCompiler             AssetCompiler
+	Now                       func() time.Time
+	projects                  []models.ProjectSpec
```
why inject them? Why not just fetch them from the repository when needed?

arinda-arif replied:
fixing this

arinda-arif replied:
changed []models.ProjectSpec to ProjectRepoFactory

kushsharma commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,181 @@
+		}
+	}
+
+	var resolvedSpecs []models.JobSpec
```
you are not really using this variable anywhere

arinda-arif replied:
fixed

kushsharma commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,181 @@
+			break
+		}
+	}
+	jobSpecMap, err := s.prepareJobSpecMap(projSpec)
```
I think we store the execution tree in db, why do we need to create the tree again from scratch?

arinda-arif replied:
no we are not storing the replay tree in db, but i missed the discussion on why we conclude on this. thoughts?

arinda-arif replied:
execution tree is now stored in db

kushsharma commented on Makefile:
```diff
@@ -27,7 +27,7 @@ pack-files:
 	@echo " > cloning protobuf from odpf/proton"
 	@rm -rf proton/
-	@git -c advice.detachedHead=false clone https://github.com/odpf/proton --depth 1 --quiet --branch main
+	@git -c advice.detachedHead=false clone https://github.com/odpf/proton --depth 1 --quiet --branch optimus-replay-status
```
I have merged the PR, you can change it to `main`

arinda-arif replied:
updated

kushsharma commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,190 @@
+	}
+	allNodes := treeNode.GetAllNodes()
+
+	stateSummary, err := s.checkInstanceState(context, projSpec, allNodes, replayRequest.Start, replayRequest.End)
```
why not use `treeNode.GetAllNodes()` directly here?

arinda-arif replied:
updated

kushsharma commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,190 @@
+			logger.I(fmt.Sprintf("marking replay as success error: %s", updateStatusErr))
+		}
+	}
+	return nil
```
I think you need to return error here `updateStatusErr `

arinda-arif replied:
added the return

arinda-arif commented:
Hi @kushsharma ,

I have made several changes, please help to check:
- Storing execution tree in DB (in replay manager), and fetch it when being used in replay worker and syncer.
- Queue to be taken by replay worker is now only represented by uuid only, since in replay worker we are fetching from db anyway.
- Not use pointers in ReplayRequest and ReplaySpec objects.
- Getting replay from repository is now preloaded with project and secret.

kushsharma commented on job/service.go:
```diff
@@ -54,7 +55,19 @@ type JobRepoFactory interface {
+	Init()
+	Replay(context.Context, models.ReplayRequest) (string, error)
+	GetReplay(uuid uuid.UUID) (models.ReplaySpec, error)
+	GetRunStatus(ctx context.Context, replaySpec models.ReplaySpec, jobSpec models.JobSpec) ([]models.JobStatus, error)
```
I see replay spec already has job spec then why are we explicitly passing it again as an argument?

arinda-arif replied:
updated to not use replaySpec and replaced with the necessary fields to avoid confusion

kushsharma commented on store/postgres/replay_repository.go:
```diff
@@ -19,57 +22,112 @@ type Replay struct {
 
+func toTreeNode(executionTree *ExecutionTree) *tree.TreeNode {
+	var jobSpec models.JobSpec
+	mapstructure.Decode(executionTree.Data, &jobSpec)
```
why are we ignoring error returned by this `Decode`?

arinda-arif replied:
refactored to not use mapstructure

kushsharma commented on store/postgres/replay_repository.go:
```diff
@@ -83,13 +141,23 @@ func (repo *replayRepository) Insert(replay *models.ReplaySpec) error {
 	var r Replay
-	if err := repo.DB.Where("id = ?", id).Find(&r).Error; err != nil {
+	if err := repo.DB.Where("id = ?", id).Preload("Job").Preload("Job.Project").
+		Preload("Job.Project.Secrets").Find(&r).Error; err != nil {
```
Can we skip loading Projects & secrets? Sorry, I can't find where we are using them?

arinda-arif replied:
we are using them in scheduler, since we need scheduler auth token and scheduler host to get dag run status. should i just load them just before calling the scheduler then? in replay manager?

arinda-arif replied:
updated to not preload projects and secrets, since in replay syncer we fetched all the projects and when getting replay, we are getting it from the user

arinda-arif commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,126 @@
+			return nil, err
+		}
+		for _, jobStatus := range jobStatusAllRuns {
+			stateSummary[jobStatus.State.String()] = stateSummary[jobStatus.State.String()] + 1
```
updated

arinda-arif commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,126 @@
+			logger.I(fmt.Sprintf("marking long running replay jobs as failed: %s", updateStatusErr))
+		}
+	}
+	return nil
```
updated

arinda-arif commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+func syncTimedOutReplay(replaySpecRepo store.ReplaySpecRepository, replaySpec models.ReplaySpec, runTimeout time.Duration) error {
+	runningTime := time.Now().Sub(replaySpec.CreatedAt)
```
updated

arinda-arif commented on job/replay_manager.go:
```diff
@@ -236,6 +131,17 @@ func (m *Manager) spawnServiceWorker() {
 
+// AirflowSyncer to sync for latest replay status
+func (m *Manager) AirflowSyncer() {
+	logger.I("start synchronizing replays...")
```
updated

arinda-arif commented on cmd/server/server.go:
```diff
@@ -79,10 +79,11 @@ func (fac *projectJobSpecRepoFactory) New(project models.ProjectSpec) store.Proj
 type replaySpecRepoRepository struct {
 	db             *gorm.DB
 	jobSpecRepoFac jobSpecRepoFactory
+	hash           models.ApplicationKey
```
updated, no longer need hash since we are not preload secrets in replay repo

arinda-arif commented on api/handler/v1/adapter_test.go:
```diff
@@ -29,6 +32,25 @@ func TestAdapter(t *testing.T) {
+		assert.Nil(t, err)
+		assert.Equal(t, replayExecutionTreeNode.JobName, "job-name")
+		assert.Equal(t, 1, len(replayExecutionTreeNode.Dependents))
+		assert.Equal(t, replayExecutionTreeNode.Dependents[0].JobName, "nested-job-name")
```
updated

arinda-arif commented on api/handler/v1/runtime.go:
```diff
@@ -803,47 +806,85 @@ func (sv *RuntimeServiceServer) Replay(ctx context.Context, req *pb.ReplayReques
+	uuid, err := uuid.Parse(req.Id)
 	if err != nil {
-		return nil, status.Errorf(codes.NotFound, "%s: project %s not found", err.Error(), req.GetProjectName())
+		return nil, status.Errorf(codes.Internal, "error while getting replay: %v", err)
```
updated

kushsharma commented on job/replay_validator.go:
```diff
@@ -0,0 +1,133 @@
+	ReplayStatusToValidate = []string{models.ReplayStatusInProgress, models.ReplayStatusAccepted, models.ReplayStatusReplayed}
+)
+
+type ReplayValidator interface {
+	Validate(context.Context, store.ReplaySpecRepository, models.ReplayRequest, *tree.TreeNode) error
+}
```
We should move this to `replay_manager.go`

arinda-arif replied:
moved

kushsharma commented on job/replay_validator.go:
```diff
@@ -0,0 +1,133 @@
+			return err
+		}
+		activeNodes := activeTree.GetAllNodes()
+		return checkAnyConflictedDags(activeNodes, reqReplayNodes)
```
This for loop will run only once instead of iterating over all `activeReplaySpecs`. Is this expected?

arinda-arif replied:
have updated this + added a test case to check this

arinda-arif commented on job/replay.go:
```diff
@@ -16,37 +18,40 @@ const (
+	replayRequest.JobSpecMap = jobSpecMap
 
-	rootInstance, err := prepareTree(replayRequest)
+	rootInstance, err := prepareReplayExecutionTree(replayRequest)
```
updated

arinda-arif commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,136 @@
+
+type ReplaySyncer interface {
+	Sync(context.Context, time.Duration) error
+}
```
moved

arinda-arif commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,136 @@
+			Type:    ReplayRunTimeout,
+			Message: fmt.Sprintf("replay has been running since %s", replaySpec.CreatedAt.UTC().Format(TimestampLogFormat)),
+		}); updateStatusErr != nil {
+			logger.I(fmt.Sprintf("marking long running replay jobs as failed: %s", updateStatusErr))
```
updated

arinda-arif commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,136 @@
+			Type:    models.ReplayStatusFailed,
+			Message: ReplayMessageFailed,
+		}); updateStatusErr != nil {
+			logger.I(fmt.Sprintf("marking replay as failed error: %s", updateStatusErr))
```
updated

arinda-arif commented on job/replay_syncer.go:
```diff
@@ -0,0 +1,136 @@
+			Type:    models.ReplayStatusSuccess,
+			Message: ReplayMessageSuccess,
+		}); updateStatusErr != nil {
+			logger.I(fmt.Sprintf("marking replay as success error: %s", updateStatusErr))
```
updated

arinda-arif commented:
Hi @kushsharma, I have updated as the feedback. what I also added:
- start & finish syncer message log are now all in Debug mode (logger.D)
- added log info when a replay has been marked as success in syncer
- Process function in Replay Mock is now waiting for 200ms instead of 100ms, due to intermittent test error

kushsharma commented:
I have removed that sleep completely and emulating it with channels.
